Error after compiling some quoted forms

Hi guys

Sorry, I found a bug on the code I submitted yesterday for fix #606 .

I had tested the code with ansi-tests and abcl-tests as usual and even added some more unit tests in the PR but after I started to work
today on one of my personal projects got an error.

I isolated the code, and after some debugging I have a patch that I think solves the issue without compromise the previous fix.

I'll send the PR in some minutes.

This is the error:

```
alejandrozf@ThinkBook-15-G2-ITL:~/projects/abcl$ cat ~/projects/lisp-experiments/abcl-issue-606.lisp 
(defmacro cp5e (codelist)
  `(list '(list) ,codelist))
alejandrozf@ThinkBook-15-G2-ITL:~/projects/abcl$ abcl
Armed Bear Common Lisp 1.9.3-dev
Java 11.0.20.1 Ubuntu
OpenJDK 64-Bit Server VM
Low-level initialization completed in 0.12 seconds.
Startup completed in 0.714 seconds.
Loading /home/alejandrozf/.abclrc completed in 10.363 seconds.
Type ":help" for a list of available commands.
CL-USER(1): (load (compile-file "~/projects/lisp-experiments/abcl-issue-606.lisp"))
; Compiling /home/alejandrozf/projects/lisp-experiments/abcl-issue-606.lisp ...
; (DEFMACRO CP5E ...)
; Wrote /home/alejandrozf/projects/lisp-experiments/abcl-issue-606.abcl (0.05 seconds)
java.lang.ClassFormatError: Truncated class file
	at java.base/java.lang.ClassLoader.defineClass1(Native Method)
	at java.base/java.lang.ClassLoader.defineClass(ClassLoader.java:1022)
	at java.base/java.lang.ClassLoader.defineClass(ClassLoader.java:883)
	at org.armedbear.lisp.JavaClassLoader.defineLispClass(JavaClassLoader.java:186)
	at org.armedbear.lisp.FaslClassLoader.findClass(FaslClassLoader.java:96)
	at org.armedbear.lisp.FaslClassLoader.loadClass(FaslClassLoader.java:73)
	at org.armedbear.lisp.JavaClassLoader.loadClass(JavaClassLoader.java:60)
	at org.armedbear.lisp.FaslClassLoader.loadFunction(FaslClassLoader.java:129)
	at org.armedbear.lisp.FaslClassLoader$pf_get_fasl_function.execute(FaslClassLoader.java:165)
	at org.armedbear.lisp.LispThread.execute(LispThread.java:930)
	at org.armedbear.lisp.Lisp.evalCall(Lisp.java:667)
	at org.armedbear.lisp.Lisp.eval(Lisp.java:607)
	at org.armedbear.lisp.Lisp.evalCall(Lisp.java:663)
	at org.armedbear.lisp.Lisp.eval(Lisp.java:607)
	at org.armedbear.lisp.Lisp.evalCall(Lisp.java:663)
	at org.armedbear.lisp.Lisp.eval(Lisp.java:607)
	at org.armedbear.lisp.Lisp.progn(Lisp.java:828)
	at org.armedbear.lisp.SpecialOperators$sf_progn.execute(SpecialOperators.java:284)
	at org.armedbear.lisp.Lisp.eval(Lisp.java:583)
	at org.armedbear.lisp.Load.faslLoadStream(Load.java:676)
	at org.armedbear.lisp.Load$init_fasl.execute(Load.java:457)
	at org.armedbear.lisp.LispThread.execute(LispThread.java:930)
	at org.armedbear.lisp.Lisp.evalCall(Lisp.java:667)
	at org.armedbear.lisp.Lisp.eval(Lisp.java:607)
	at org.armedbear.lisp.Load.loadStream(Load.java:638)
	at org.armedbear.lisp.Load.loadFileFromStream(Load.java:606)
	at org.armedbear.lisp.Load.load(Load.java:213)
	at org.armedbear.lisp.Load.load(Load.java:754)
	at org.armedbear.lisp.Load$_load.execute(Load.java:702)
	at org.armedbear.lisp.Symbol.execute(Symbol.java:854)
	at org.armedbear.lisp.LispThread.execute(LispThread.java:998)
	at org.armedbear.lisp.load_1.execute(load.lisp:33)
	at org.armedbear.lisp.CompiledClosure.execute(CompiledClosure.java:98)
	at org.armedbear.lisp.LispThread.execute(LispThread.java:910)
	at org.armedbear.lisp.Lisp.evalCall(Lisp.java:656)
	at org.armedbear.lisp.Lisp.eval(Lisp.java:607)
	at org.armedbear.lisp.Primitives$pf__eval.execute(Primitives.java:346)
	at org.armedbear.lisp.LispThread.execute(LispThread.java:910)
	at org.armedbear.lisp.Lisp.evalCall(Lisp.java:656)
	at org.armedbear.lisp.Lisp.eval(Lisp.java:607)
	at org.armedbear.lisp.Lisp.progn(Lisp.java:828)
	at org.armedbear.lisp.Primitives$sf_block.execute(Primitives.java:3760)
	at org.armedbear.lisp.Lisp.eval(Lisp.java:583)
	at org.armedbear.lisp.Lisp.progn(Lisp.java:828)
	at org.armedbear.lisp.Closure.execute(Closure.java:220)
	at org.armedbear.lisp.Closure.execute(Closure.java:148)
	at org.armedbear.lisp.LispThread.execute(LispThread.java:910)
	at org.armedbear.lisp.Lisp$1.execute(Lisp.java:288)
	at org.armedbear.lisp.Symbol.execute(Symbol.java:805)
	at org.armedbear.lisp.LispThread.execute(LispThread.java:910)
	at org.armedbear.lisp.top_level_47.execute(top-level.lisp:407)
	at org.armedbear.lisp.CompiledClosure.execute(CompiledClosure.java:89)
	at org.armedbear.lisp.Symbol.execute(Symbol.java:795)
	at org.armedbear.lisp.LispThread.execute(LispThread.java:892)
	at org.armedbear.lisp.top_level_48.execute(top-level.lisp:415)
	at org.armedbear.lisp.LispThread.execute(LispThread.java:892)
	at org.armedbear.lisp.Interpreter.run(Interpreter.java:361)
	at org.armedbear.lisp.Main$1.run(Main.java:48)
	at java.base/java.lang.Thread.run(Thread.java:829)
java.lang.ClassNotFoundException: Function class not found: org.armedbear.lisp.abcl_issue_606_1
	at org.armedbear.lisp.FaslClassLoader.findClass(FaslClassLoader.java:100)
	at org.armedbear.lisp.FaslClassLoader.loadClass(FaslClassLoader.java:73)
	at org.armedbear.lisp.JavaClassLoader.loadClass(JavaClassLoader.java:60)
	at org.armedbear.lisp.FaslClassLoader.loadFunction(FaslClassLoader.java:129)
	at org.armedbear.lisp.FaslClassLoader$pf_get_fasl_function.execute(FaslClassLoader.java:165)
	at org.armedbear.lisp.LispThread.execute(LispThread.java:930)
	at org.armedbear.lisp.Lisp.evalCall(Lisp.java:667)
	at org.armedbear.lisp.Lisp.eval(Lisp.java:607)
	at org.armedbear.lisp.Lisp.evalCall(Lisp.java:663)
	at org.armedbear.lisp.Lisp.eval(Lisp.java:607)
	at org.armedbear.lisp.Lisp.evalCall(Lisp.java:663)
	at org.armedbear.lisp.Lisp.eval(Lisp.java:607)
	at org.armedbear.lisp.Lisp.progn(Lisp.java:828)
	at org.armedbear.lisp.SpecialOperators$sf_progn.execute(SpecialOperators.java:284)
	at org.armedbear.lisp.Lisp.eval(Lisp.java:583)
	at org.armedbear.lisp.Load.faslLoadStream(Load.java:676)
	at org.armedbear.lisp.Load$init_fasl.execute(Load.java:457)
	at org.armedbear.lisp.LispThread.execute(LispThread.java:930)
	at org.armedbear.lisp.Lisp.evalCall(Lisp.java:667)
	at org.armedbear.lisp.Lisp.eval(Lisp.java:607)
	at org.armedbear.lisp.Load.loadStream(Load.java:638)
	at org.armedbear.lisp.Load.loadFileFromStream(Load.java:606)
	at org.armedbear.lisp.Load.load(Load.java:213)
	at org.armedbear.lisp.Load.load(Load.java:754)
	at org.armedbear.lisp.Load$_load.execute(Load.java:702)
	at org.armedbear.lisp.Symbol.execute(Symbol.java:854)
	at org.armedbear.lisp.LispThread.execute(LispThread.java:998)
	at org.armedbear.lisp.load_1.execute(load.lisp:33)
	at org.armedbear.lisp.CompiledClosure.execute(CompiledClosure.java:98)
	at org.armedbear.lisp.LispThread.execute(LispThread.java:910)
	at org.armedbear.lisp.Lisp.evalCall(Lisp.java:656)
	at org.armedbear.lisp.Lisp.eval(Lisp.java:607)
	at org.armedbear.lisp.Primitives$pf__eval.execute(Primitives.java:346)
	at org.armedbear.lisp.LispThread.execute(LispThread.java:910)
	at org.armedbear.lisp.Lisp.evalCall(Lisp.java:656)
	at org.armedbear.lisp.Lisp.eval(Lisp.java:607)
	at org.armedbear.lisp.Lisp.progn(Lisp.java:828)
	at org.armedbear.lisp.Primitives$sf_block.execute(Primitives.java:3760)
	at org.armedbear.lisp.Lisp.eval(Lisp.java:583)
	at org.armedbear.lisp.Lisp.progn(Lisp.java:828)
	at org.armedbear.lisp.Closure.execute(Closure.java:220)
	at org.armedbear.lisp.Closure.execute(Closure.java:148)
	at org.armedbear.lisp.LispThread.execute(LispThread.java:910)
	at org.armedbear.lisp.Lisp$1.execute(Lisp.java:288)
	at org.armedbear.lisp.Symbol.execute(Symbol.java:805)
	at org.armedbear.lisp.LispThread.execute(LispThread.java:910)
	at org.armedbear.lisp.top_level_47.execute(top-level.lisp:407)
	at org.armedbear.lisp.CompiledClosure.execute(CompiledClosure.java:89)
	at org.armedbear.lisp.Symbol.execute(Symbol.java:795)
	at org.armedbear.lisp.LispThread.execute(LispThread.java:892)
	at org.armedbear.lisp.top_level_48.execute(top-level.lisp:415)
	at org.armedbear.lisp.LispThread.execute(LispThread.java:892)
	at org.armedbear.lisp.Interpreter.run(Interpreter.java:361)
	at org.armedbear.lisp.Main$1.run(Main.java:48)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: java.lang.ClassFormatError: Truncated class file
	at java.base/java.lang.ClassLoader.defineClass1(Native Method)
	at java.base/java.lang.ClassLoader.defineClass(ClassLoader.java:1022)
	at java.base/java.lang.ClassLoader.defineClass(ClassLoader.java:883)
	at org.armedbear.lisp.JavaClassLoader.defineLispClass(JavaClassLoader.java:186)
	at org.armedbear.lisp.FaslClassLoader.findClass(FaslClassLoader.java:96)
	... 54 more
Error loading /home/alejandrozf/projects/lisp-experiments/abcl-issue-606.abcl at line 10 (offset 686)
#<THREAD "interpreter" native {2DBD806A}>: Debugger invoked on condition of type ERROR
  Compiled function can't be loaded: org.armedbear.lisp.abcl_issue_606_1 from org.armedbear.lisp.Pathname@58dde996
Restarts:
  0: TOP-LEVEL Return to top level.
[1] CL-USER(2): 
```

--- a/src/org/armedbear/lisp/compiler-pass2.lisp
+++ b/src/org/armedbear/lisp/compiler-pass2.lisp
@@ -1301,9 +1301,8 @@ of the other types."
                        (return-from sh-str-helper nil)
                        (setf (gethash obj circularity-hashtable) t))))
                (cond
-                 ((atom obj)
-                  (when (assoc obj *externalized-objects* :test #'eq)
-                    (return-from shared-structure-p t)))
+                 ((and (atom obj) (not (eq obj 'quote)))
+                  (return-from shared-structure-p nil))
                  ((listp obj)
                   (when (assoc obj *externalized-objects* :test #'eq)
                     (return-from shared-structure-p t))

--- a/test/lisp/abcl/compiler-tests.lisp
+++ b/test/lisp/abcl/compiler-tests.lisp
@@ -578,3 +578,17 @@
       (prog1 (abcl/test/lisp::cp5d)
         (fmakunbound 'abcl/test/lisp::cp5d)))
   nil)
+
+#+abcl
+(deftest compiler.5e
+    (let ((tmpfile (ext::make-temp-file))
+          (code
+             "(defmacro cp5e (codelist)
+               `(list '(list) ,codelist))"))
+      (with-open-file (s tmpfile :direction :output)
+        (format s "~A" code))
+      (prog1
+          (load (compile-file tmpfile))
+        (delete-file tmpfile)
+        (fmakunbound 'abcl/test/lisp::cp5e)))
+  t)

